### ISSUE TITLE
Weird conflict between formatter and linter fix all

### ISSUE THREAD

**Issue by mgzenitech** [astral-sh/ruff#10624]
Here are two files:
```
#!/usr/bin/env -S poetry run python

from typing import Any
from toml import load

# @BUG: https://github.com/darrenburns/ward/issues/378
from ward import Scope, fixture  # pyright: ignore [reportUnknownVariableType]
```

and

```
#!/usr/bin/env -S poetry run python

from collections.abc import Callable
from typing import Any
from unittest.mock import MagicMock

# @BUG: https://github.com/darrenburns/ward/issues/378
from ward import Scope, fixture  # pyright: ignore [reportUnknownVariableType]
```

First one doesn't have any issues yet the second one is triggering `I001`,

.ruff.toml:
```
builtins = [
  "_"
]
cache-dir = "tmp/ruff"
preview = true
target-version = "py312"
unsafe-fixes = true

[format]
docstring-code-format = true
docstring-code-line-length = 120
line-ending = "lf"

[lint]
dummy-variable-rgx = "^_$"
extend-select = [
  "D213"
]
ignore = [
  "CPY001",
  "D212"
]
select = [
  "ALL"
]
task-tags = [
  "@BUG",
  "@HACK",
  "@TODO"
]

[lint.isort]
combine-as-imports = true
force-sort-within-sections = true
lines-after-imports = 2
no-lines-before = [
  "first-party",
  "future",
  "local-folder",
  "standard-library",
  "third-party"
]

[lint.pycodestyle]
max-doc-length = 120
max-line-length = 120

[lint.pydocstyle]
convention = "google"

[lint.pylint]
allow-magic-value-types = [
]

[lint.flake8-annotations]
allow-star-arg-any = true
```

**Comment by MichaReiser** [comment]
Thanks for reporting this bug. 

The issue here is that the formatter enforces a blank line between two imports if there's an own-line comment between them (similar to black).

```py
from unittest.mock import MagicMock
# @BUG: https://github.com/darrenburns/ward/issues/378
from ward import Scope, fixture  # pyright: ignore [reportUnknownVariableType]
```

Gets reformatted to 

```py
from unittest.mock import MagicMock

# @BUG: https://github.com/darrenburns/ward/issues/378
from ward import Scope, fixture  # pyright: ignore [reportUnknownVariableType]
```

Now, `isort` doesn't agree with the blank line because of the `no-lines-before`  configuration.

**Comment by mgzenitech** [comment]
Will formatter be updated to take into account .ruff.toml config of isort?

**Comment by MichaReiser** [comment]
I need to think in more detail about what the best solution here is. We want to avoid that the `formatter` needs to read the `isort` configuration and reimplement the same logic (the formatter doesn't know about first-party-imports etc).

**Comment by mgzenitech** [comment]
I think I've found a bunch of other issues too with formatter conflicting lint fix. Long term formatter should be compatible with linter config file...

**Comment by charliermarsh** [comment]
For what it's worth, by far the best way to ensure compatibility is to avoid highly customized formatting rules or configuration in the linter (e.g., modifying the `isort` settings). It's possible that we don't end up "fixing" those and instead just mark them as incompatible with user-facing warnings, as with other formatter-conflicting lint rules and settings (see: https://docs.astral.sh/ruff/formatter/#conflicting-lint-rules).

**Comment by ERW105** [comment]
> I need to think in more detail about what the best solution here is. We want to avoid that the `formatter` needs to read the `isort` configuration and reimplement the same logic (the formatter doesn't know about first-party-imports etc).

In the original isort, it appears that `ensure_newline_before_comments` takes priority over `no_lines_before`, such that there is always a newline before comments and the output is black-compatible. Including when a comment is before the first import of a `no_lines_before` section.

`ensure_newline_before_comments` is part of isort's black profile and I assume the behavior ruff implements implicitly. Perhaps this divergence is unintended? I don't think we need the formatter to be involved here.

Super impressed with ruff and would love to keep using `no-lines-before`!

**Comment by MichaReiser** [comment]
Nice find @ERW105 Yes, implementing the behavior of `ensure_newline_before_comments` -- at least if `no_lines_before = 0` (without adding an option) is probably the solution here, although it is a breaking change.
